torch/csrc/jit/codegen/cuda/lower_index.cpp

```diff
@@ -921,6 +921,103 @@ std::vector<kir::Allocate*> IndexLowering::allocateWelfordWorkBuffer(
   return work_buffers;
 }
 
+namespace {
+
+// Returns true if a GroupedWelfordOp op is eligible for using the
+// outer-optimized grouped welford runtime function
+bool canUseOuterOptRuntimeKernel(const GroupedWelfordOp* grouped_wop) {
+  const auto out_tv = ir_utils::getTvOutput(grouped_wop);
+  const auto out_domain = out_tv->domain();
+
+  if (!out_domain->hasGridReduction()) {
+    return false;
+  }
+
+  // TIDx and BIDx must be used for non-reduction domains. TIDy and
+  // BIDy must be used for reduction domains.
+  ParallelTypeBitmap used_pts;
+  for (auto leaf_id : out_domain->domain()) {
+    auto pt = leaf_id->getParallelType();
+    if (isParallelTypeThread(pt)) {
+      used_pts.set(pt);
+      if ((leaf_id->isReduction() &&
+           (pt == ParallelType::BIDy || pt == ParallelType::TIDy)) ||
+          (leaf_id->getIterType() == IterType::Iteration &&
+           (pt == ParallelType::BIDx || pt == ParallelType::TIDx))) {
+        // valid pattern
+        continue;
+      } else {
+        return false;
+      }
+    }
+  }
+
+  ParallelTypeBitmap valid_pt_map;
+  valid_pt_map.set(ParallelType::BIDx);
+  valid_pt_map.set(ParallelType::BIDy);
+  valid_pt_map.set(ParallelType::TIDx);
+  valid_pt_map.set(ParallelType::TIDy);
+  if (used_pts != valid_pt_map) {
+    return false;
+  }
+
+  // TIDx and TIDy must be static constant
+  const auto& par_dim_map = GpuLower::current()->parallelDimensionMap();
+  auto tidx = par_dim_map.get(ParallelType::TIDx)->getInt();
+  auto tidy = par_dim_map.get(ParallelType::TIDy)->getInt();
+  if (!tidx.has_value() || !tidy.has_value()) {
+    return false;
+  }
+
+  // Warp reduction along threadIdx.y is a key factor for the
+  // outer-optimized kernel. The larger (32 / blockDim.x) is, the more
+  // effective. It shouldn't give any perf benefit when blockDim.x >=
+  // 32 as there's no warp reduction. blockDim.x == 16 is not
+  // preferable, but still would be better than the default
+  // implementation. blockDim.x == 8 is preferred.
+  if (tidx > 16) {
+    return false;
+  }
+
+  int num_grouped_iterations = 1;
+  for (auto axis : out_domain->domain()) {
+    if (axis->getParallelType() == ParallelType::Group) {
+      num_grouped_iterations *= axis->extent()->getInt().value();
```

Comment:
should probably still just use `evaluateInt` or assert the value is constant so you can just pull the value from it.